utils/fork-tree/src/lib.rs

```diff
@@ -267,6 +189,57 @@ where
 		self.node_iter().map(|node| (&node.hash, &node.number, &node.data))
 	}
 
+	/// Map fork tree into values of new types.
+	pub fn map<VT, F>(self, f: &mut F) -> ForkTree<H, N, VT>
+	where
+		F: FnMut(&H, &N, V) -> VT,
+	{
+		let mut new_tree = ForkTree {
+			roots: Default::default(),
+			best_finalized_number: self.best_finalized_number,
+		};
+
+		let mut old_children = self.roots;
+		let mut new_children = &mut new_tree.roots;
+
+		let mut stack = vec![];
+
+		loop {
+			let mut tmp = vec![];
+			for node in old_children {
+				let data = f(&node.hash, &node.number, node.data);
+				new_children.push(Node {
+					data,
+					hash: node.hash,
+					number: node.number,
+					children: Default::default(),
+				});
+
+				tmp.push(node.children);
+			}
+
+			// SAFETY: borrow checker doesn't allow mutable references to multiple parts of
+			// the tree. Usage or raw pointers is safe as far as we take care to handle
+			// only pointers to data that is not subject to relocation.
+			//
+			// This should be performed after the node children have a stable addresses,
+			// i.e. vector buffer not subject to relocation due to some element insertion.
+			for (i, old_children) in tmp.into_iter().enumerate() {
+				stack.push((old_children, &mut new_children[i].children as *mut _));
+			}
+			// Process the next children lists.
+			match stack.pop() {
+				Some((old, new_ptr)) => {
+					old_children = old;
+					new_children = unsafe { &mut *new_ptr };
+				},
+				None => break,
+			};
+		}
+
+		new_tree
+	}
```

Comment:
Alternative implementation I've quickly hacked up, without the `unsafe`:

```rust
        let mut queue: Vec<_> = self.roots.into_iter().rev().map(|node| (usize::MAX, node)).collect();
        let mut next_queue = Vec::new();
        let mut output = Vec::new();

        while !queue.is_empty() {
            for (parent_index, node) in queue.drain(..) {
                let new_data = f(&node.hash, &node.number, node.data);
                let new_node = Node {
                    hash: node.hash,
                    number: node.number,
                    data: new_data,
                    children: Vec::with_capacity(node.children.len())
                };

                let node_id = output.len();
                output.push((parent_index, new_node));

                for child in node.children.into_iter().rev() {
                    next_queue.push((node_id, child));
                }
            }

            std::mem::swap(&mut queue, &mut next_queue);
        }

        let mut roots = Vec::new();
        while let Some((parent_index, new_node)) = output.pop() {
            if parent_index == usize::MAX {
                roots.push(new_node);
            } else {
                output[parent_index].1.children.push(new_node);
            }
        }

        ForkTree {
            roots,
            best_finalized_number: self.best_finalized_number
        }
```

Might or might not be better. Passes the tests though. (Although we're missing a test that makes sure that the roots are outputted in the same order.)